Add Nice G.O. integration

IceBotYT opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add the Nice G.O. integration.
Users coming from Linear Garage Door will need to download the Nice G.O. app and migrate their account first and update the firmware of their garage door(s) in order to use this integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #122259 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/34003

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

IceBotYT committed:
Convert Linear Garage Door to Nice G.O.

IceBotYT committed:
Remove useless fixtures

IceBotYT committed:
Update manifest (now cloud push! 🎉)

IceBotYT committed:
Fix entry unload

IceBotYT commented:
I don't believe the mypy errors are related to this PR. Correct me if I'm wrong though

joostlek commented on homeassistant/components/linear_garage_door/__init__.py:
```diff
@@ -1,25 +1,45 @@
+    coordinator = NiceGOUpdateCoordinator(hass)
 
     await coordinator.async_config_entry_first_refresh()
+    entry.runtime_data = coordinator
```
We should extend the typing of the config entry to make it type safe

joostlek commented on homeassistant/components/linear_garage_door/config_flow.py:
```diff
@@ -80,6 +72,9 @@ async def async_step_user(
         if user_input is None:
             return self.async_show_form(step_id="user", data_schema=data_schema)
 
+        await self.async_set_unique_id(user_input[CONF_EMAIL])
```
Is the email unique?

IceBotYT replied:
I believe so yeah, I could call `.lower()` or `.upper()` on it if you'd like

joostlek commented on homeassistant/components/linear_garage_door/diagnostics.py:
```diff
@@ -10,17 +10,17 @@
 ) -> dict[str, Any]:
     """Return diagnostics for a config entry."""
-    coordinator: LinearUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    coordinator: NiceGOUpdateCoordinator = entry.runtime_data
```
```suggestion
    coordinator = entry.runtime_data
```
The moment you have the extended config entry type, this isn't needed anymore

IceBotYT committed:
Extend config entry type

IceBotYT committed:
Fix circular import

joostlek commented:
What dependencies exactly?

IceBotYT commented:
Looks to be `boto3`

IceBotYT committed:
Bump nice-go (hopefully fix dep conflict)

IceBotYT committed:
Bump nice-go (moves type stubs to dev deps)

IceBotYT commented:
Fixed the `mypy` error. I moved the type stubs to the dev group and not have them install here in HA. The test failure though, I'm sure it's not related to this PR, as I've seen it error here as well https://github.com/home-assistant/core/actions/runs/10152673631/job/28075712980

IceBotYT commented:
@coderabbitai review

IceBotYT committed:
Remove lingering mentions of Linear

mib1185 commented:
> Users will need to migrate their account and update the firmware of their garage doors to continue to use this integration.

I think this needs to be elaborated a bit more detailed or at least described in detail in the integration docs.
Is there already an ETA when the old integration/api/lib will not work anymore and users will be confronted with a not-functional integration?

IceBotYT commented:
> I think this needs to be elaborated a bit more detailed or at least described in detail in the integration docs.

The Linear app recently started showing a pop-up upon opening informing users that the app will be discontinued in the coming weeks and to move to the new app (Nice G.O.)

![image](https://github.com/user-attachments/assets/44eaa9be-297a-44b2-a90e-ced45be7d6e1)

Upon downloading the new app, devices and other info is migrated automatically, and that causes the old app to stop working. Inside the new app, you won't be able to use your garage doors until you install a firmware update on the doors, only showing an "update needed" icon and a "start update" button.

> Is there already an ETA when the old integration/api/lib will not work anymore and users will be confronted with a not-functional integration?

Only ETA I have is from the app, only mentioning "the next few weeks". See image above.

mib1185 commented:
as discussed on [discord](https://discord.com/channels/330944238910963714/1222192972519706798/1268356989436825640) this should go into current beta, since users might end up in a broken integration

IceBotYT committed:
Add nice-go as logger

edenhaus commented:
> as discussed on [discord](https://discord.com/channels/330944238910963714/1222192972519706798/1268356989436825640) this should go into current beta, since users might end up in a broken integration

It will be a breaking change for users who haven't upgraded yet, which is also not nice. Would it be possible to support both apps until the old one shuts down?
The advantage to supporting both is that we don't force users to migrate on release day.

IceBotYT commented:
I could potentially add legacy support, but it may not be worth it in the long term because the app is shutting down in only the next few weeks, and I won't have any way of testing it with real devices myself, because upgrading to the new app stops the old app from working.

DocBrownInSpace commented:
> I could potentially add legacy support, but it may not be worth it in the long term because the app is shutting down in only the next few weeks, and I won't have any way of testing it with real devices myself, because upgrading to the new app stops the old app from working.

IMHO, it’s not worth the effort to support both APIs.  Since anyone broken by the update, simply has to migrate to the new app.  If they choose not to migrate, they still end up in a broken state, as the old app will be retired soon.  One way or the other, everyone has to migrate.

joostlek commented:
Marked draft as we thought that because of the hard migration they are executing, it's maybe easier to just make this a new integration instead. It won't invalidate the efforts, as only the domain has to change (iirc)

IceBotYT committed:
Convert nice_go into a new integration and restore linear_garage_door

IceBotYT committed:
Add missing new lines to snapshots

joostlek commented on homeassistant/components/nice_go/__init__.py:
```diff
@@ -0,0 +1,43 @@
+from .coordinator import NiceGOUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+PLATFORMS: list[Platform] = [Platform.COVER, Platform.LIGHT]
```
Let's still try to use the normal rules and only add the cover in this PR

joostlek commented on homeassistant/components/nice_go/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        data_schema = vol.Schema(STEP_USER_DATA_SCHEMA)
```
why isnt the constant stored as vol.Schema?

joostlek commented on homeassistant/components/nice_go/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+        """Handle the initial step."""
+        data_schema = vol.Schema(STEP_USER_DATA_SCHEMA)
+
+        if user_input is None:
+            return self.async_show_form(step_id="user", data_schema=data_schema)
+
+        await self.async_set_unique_id(user_input[CONF_EMAIL])
+        self._abort_if_unique_id_configured()
+
+        errors = {}
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except InvalidAuth:
```
I'd invert it 
```py
if user_input is not None:
    ....
return self.async_show_form
```
This way you don't need multiple forms

joostlek commented on homeassistant/components/nice_go/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+            errors["base"] = "unknown"
+        else:
+            # Check if we are reauthenticating
+            if self._reauth_entry is not None:
+                self.hass.config_entries.async_update_entry(
+                    self._reauth_entry,
+                    data=info,
+                )
+                await self.hass.config_entries.async_reload(self._reauth_entry.entry_id)
+                return self.async_abort(reason="reauth_successful")
```
Let's leave reauth for a followup

joostlek commented on homeassistant/components/nice_go/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+        errors = {}
+
+        try:
+            info = await validate_input(self.hass, user_input)
```
This function can be inlined imo

joostlek commented on homeassistant/components/nice_go/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+    async def _async_update_data(self) -> dict[str, NiceGODevice]:
+        """Fetch data from Nice G.O. API."""
+        async with asyncio.timeout(10):
+            if self.api.id_token:
+                # Already authenticated, returning the same data to avoid overwriting data from the websocket
+                return self.data
```
So if I understand correctly the whole setup function is only for setting up for the first time? Use `_async_setup` that's a new feature we added this release

joostlek commented on homeassistant/components/nice_go/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+        parsed_data = await self._parse_barrier(
+            BarrierState(
+                deviceId=raw_data["deviceId"],
+                desired=json.loads(raw_data["desired"]),
+                reported=json.loads(raw_data["reported"]),
```
Do they encode json into json?

IceBotYT replied:
Yeah, it's weird

joostlek commented on homeassistant/components/nice_go/diagnostics.py:
Followup

IceBotYT committed:
Fixes suggested by @joostlek

IceBotYT committed:
More fixes

joostlek committed:
Fixes

joostlek committed:
Fixes

frenck commented:
Removed the milestone tag, as this is a new integration.

IceBotYT committed:
Fix coordinator tests

IceBotYT committed:
Move coordinator tests

IceBotYT committed:
Move test_no_connection_state from test_cover to test_init
